events/events.go

```diff
@@ -0,0 +1,379 @@
+// objects information.
+type State string
+
+const (
+	// StatePending represents the event information is pending expansion. That's
+	// done as a worker job. Users don't have access while an event is pending.
+	StatePending State = "pending"
+
+	// StateTracking represents the event is sending information to a 3rd-party
+	// analytics. Users can access an event while is tracking state.
+	StateTracking State = "tracking"
+
+	// StateDone represents the event information expansion is done. The event
+	// is immutable going forward. Users can access a done event.
+	StateDone State = "done"
+)
+
+// SourceType represents where the request came from.
+type SourceType string
+
+const (
+	// SourceDashboard is a request coming from the dashboard
+	SourceDashboard SourceType = "dashboard"
+
+	// SourceCLI is a request coming from the cli
+	SourceCLI SourceType = "cli"
+
+	// SourceSystem is an internal request
+	SourceSystem SourceType = "system"
+)
+
+// Event represents meaningful activities performed on the system.
+type Event struct {
+	ID            manifold.ID `json:"id"`
+	StructType    string      `json:"type"`
+	StructVersion int         `json:"version"`
+	State         State       `json:"state"`
+	Body          Body        `json:"body"`
+}
+
+// GetID returns the ID associated with this event
+func (e *Event) GetID() manifold.ID { return e.ID }
+
+// Version returns the data structure version of this event
+func (e *Event) Version() int { return e.StructVersion }
+
+// Type returns the idtype object for this struct type
+func (e *Event) Type() idtype.Type { return idtype.ActivityEvent }
+
+// Validate returns whether or not the given Event is valid
+func (e *Event) Validate(v interface{}) error {
+	if err := e.ID.Validate(v); err != nil {
+		return err
+	}
+
+	if e.Version() != 1 {
+		return manifold.NewError(merrors.BadRequestError, "Expected version to be 1")
+	}
+
+	return e.Body.Validate(v)
+}
+
+// StateType returns the event state type as a string
+func (e *Event) StateType() string {
+	return string(e.Body.Type())
+}
+
+// GetState returns the event's state
+func (e *Event) GetState() string {
+	return string(e.State)
+}
+
+// SetState sets the event's state
+func (e *Event) SetState(state string) {
+	e.State = State(state)
+}
+
+// SetUpdatedAt sets the event's updated at time to the current time.
+func (e *Event) SetUpdatedAt() {
+	e.Body.SetUpdatedAt()
+}
+
+type outEvent struct {
+	ID            manifold.ID     `json:"id"`
+	StructType    string          `json:"type"`
+	StructVersion int             `json:"version"`
+	Body          json.RawMessage `json:"body"`
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface for an event
+func (e *Event) UnmarshalJSON(b []byte) error {
+	t := outEvent{}
+	err := json.Unmarshal(b, &t)
+	if err != nil {
+		return err
+	}
+
+	v := BaseBody{}
+	err = json.Unmarshal(t.Body, &v)
+	if err != nil {
+		return err
+	}
+
+	var body Body
+	switch v.Type() {
+	case TypeOperationProvisioned:
```

Comment:
You have 2 other types defined, shouldn't they be handled here?

Reply:
I know you don't have the struct defined, but should they at least be allowed?

Reply:
I removed the other types from the initial PR. So we can ship the provision first, then I will re-add the other events after.

Reply:
kk :+1: